File: native-app/Sources/OutputManager.swift
import Foundation
import AppKit
import CoreGraphics

class OutputManager {
    let baseDir: String
    var currentDateFolder: String?

    private let fileManager = FileManager.default
    private let dateFormatter = ISO8601DateFormatter()

    private var ndjsonHandle: FileHandle?
    private var csvHandle: FileHandle?
    private var csvHeaderWritten = false

    init(baseDir: String) {
        self.baseDir = baseDir
        dateFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    }

    deinit {
        closeLogFiles()
    }

    /// Creates today's output folder (YYYY-MM-DD)
    func createTodayFolder() {
        let today = getTodayFolderName()
        let folderPath = (baseDir as NSString).appendingPathComponent(today)

        currentDateFolder = folderPath

        // Create directory if it doesn't exist
        try? fileManager.createDirectory(atPath: folderPath, withIntermediateDirectories: true, attributes: nil)

        // Open log files
        openLogFiles()
    }

    /// Gets today's folder name in YYYY-MM-DD format (UTC)
    private func getTodayFolderName() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.timeZone = TimeZone(identifier: "UTC")
        return formatter.string(from: Date())
    }

    /// Opens log files for writing
    private func openLogFiles() {
        guard let folder = currentDateFolder else { return }

        let ndjsonPath = (folder as NSString).appendingPathComponent("clicks.ndjson")
        let csvPath = (folder as NSString).appendingPathComponent("clicks.csv")

        // Create files if they don't exist
        if !fileManager.fileExists(atPath: ndjsonPath) {
            fileManager.createFile(atPath: ndjsonPath, contents: nil, attributes: nil)
        }

        if !fileManager.fileExists(atPath: csvPath) {
            fileManager.createFile(atPath: csvPath, contents: nil, attributes: nil)
            csvHeaderWritten = false
        } else {
            csvHeaderWritten = true // Assume header exists if file exists
        }

        // Open file handles
        ndjsonHandle = FileHandle(forWritingAtPath: ndjsonPath)
        csvHandle = FileHandle(forWritingAtPath: csvPath)

        // Seek to end
        if #available(macOS 10.15.4, *) {
            try? ndjsonHandle?.seekToEnd()
            try? csvHandle?.seekToEnd()
        } else {
            ndjsonHandle?.seekToEndOfFile()
            csvHandle?.seekToEndOfFile()
        }

        // Write CSV header if needed
        if !csvHeaderWritten, let handle = csvHandle {
            let header = ClickEvent.csvHeader + "\n"
            if let data = header.data(using: .utf8) {
                if #available(macOS 10.15.4, *) {
                    try? handle.write(contentsOf: data)
                } else {
                    handle.write(data)
                }
                csvHeaderWritten = true
            }
        }
    }

    /// Closes log files
    private func closeLogFiles() {
        if #available(macOS 10.15, *) {
            try? ndjsonHandle?.close()
            try? csvHandle?.close()
        } else {
            ndjsonHandle?.closeFile()
            csvHandle?.closeFile()
        }
    }

    /// Saves a screenshot with UTC timestamp filename
    func saveScreenshot(_ image: CGImage, displayID: CGDirectDisplayID) {
        // Check if we need to create a new folder (date changed)
        let todayFolder = getTodayFolderName()
        let currentFolder = (currentDateFolder as NSString?)?.lastPathComponent

        if todayFolder != currentFolder {
            closeLogFiles()
            createTodayFolder()
        }

        guard let folder = currentDateFolder else { return }

        // Generate filename: YYYY-MM-DDTHH-MM-SS.mmmZ.png
        let timestamp = Date()
        let filename = getScreenshotFilename(timestamp, displayID: displayID)
        let filePath = (folder as NSString).appendingPathComponent(filename)

        // Convert CGImage to PNG and save
        if let pngData = image.pngData() {
            try? pngData.write(to: URL(fileURLWithPath: filePath))
        }
    }

    /// Generates screenshot filename with ISO-8601 format
    private func getScreenshotFilename(_ date: Date, displayID: CGDirectDisplayID) -> String {
        let isoString = dateFormatter.string(from: date)

        // Format: YYYY-MM-DDTHH-MM-SS.mmmZ.png
        // ISO8601 gives us: 2025-11-11T14:30:15.123Z
        // We need to replace colons with dashes for filesystem compatibility
        let formatted = isoString.replacingOccurrences(of: ":", with: "-")

        // Add display ID suffix if multiple displays
        let displaySuffix = displayID == CGMainDisplayID() ? "" : "-display\(displayID)"

        return "\(formatted)\(displaySuffix).png"
    }

    /// Logs a click event to both NDJSON and CSV files
    func logClick(_ event: ClickEvent) {
        // Check if we need to create a new folder
        let todayFolder = getTodayFolderName()
        let currentFolder = (currentDateFolder as NSString?)?.lastPathComponent

        if todayFolder != currentFolder {
            closeLogFiles()
            createTodayFolder()
        }

        // Write to NDJSON
        if let handle = ndjsonHandle {
            let ndjson = event.toNDJSON() + "\n"
            if let data = ndjson.data(using: .utf8) {
                if #available(macOS 10.15.4, *) {
                    try? handle.write(contentsOf: data)
                } else {
                    handle.write(data)
                }
            }
        }

        // Write to CSV
        if let handle = csvHandle {
            let csv = event.toCSVRow() + "\n"
            if let data = csv.data(using: .utf8) {
                if #available(macOS 10.15.4, *) {
                    try? handle.write(contentsOf: data)
                } else {
                    handle.write(data)
                }
            }
        }

        // Flush to disk
        if #available(macOS 10.15, *) {
            try? ndjsonHandle?.synchronize()
            try? csvHandle?.synchronize()
        } else {
            ndjsonHandle?.synchronizeFile()
            csvHandle?.synchronizeFile()
        }
    }
}

File: native-app-windows/OutputManager.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.Globalization;
using System.IO;
using System.Text;
using Newtonsoft.Json;

namespace DesktopCapture
{
    class OutputManager
    {
        private readonly string baseDir;
        private string? currentDateFolder;

        private StreamWriter? ndjsonWriter;
        private StreamWriter? csvWriter;
        private bool csvHeaderWritten = false;

        public OutputManager(string baseDir)
        {
            this.baseDir = baseDir;
        }

        /// <summary>
        /// Creates today's output folder (YYYY-MM-DD in UTC)
        /// </summary>
        public void CreateTodayFolder()
        {
            string today = GetTodayFolderName();
            string folderPath = Path.Combine(baseDir, today);

            currentDateFolder = folderPath;

            // Create directory if it doesn't exist
            Directory.CreateDirectory(folderPath);

            // Open log files
            OpenLogFiles();
        }

        /// <summary>
        /// Gets today's folder name in YYYY-MM-DD format (UTC)
        /// </summary>
        private string GetTodayFolderName()
        {
            return DateTime.UtcNow.ToString("yyyy-MM-dd");
        }

        /// <summary>
        /// Opens log files for writing
        /// </summary>
        private void OpenLogFiles()
        {
            if (currentDateFolder == null)
                return;

            string ndjsonPath = Path.Combine(currentDateFolder, "clicks.ndjson");
            string csvPath = Path.Combine(currentDateFolder, "clicks.csv");

            // Check if CSV file already exists (header already written)
            csvHeaderWritten = File.Exists(csvPath);

            // Open file streams for appending
            ndjsonWriter = new StreamWriter(ndjsonPath, append: true, Encoding.UTF8);
            csvWriter = new StreamWriter(csvPath, append: true, Encoding.UTF8);

            // Set auto-flush for immediate writes
            ndjsonWriter.AutoFlush = true;
            csvWriter.AutoFlush = true;

            // Write CSV header if needed
            if (!csvHeaderWritten)
            {
                csvWriter.WriteLine("ts,x,y,app,window_title,role,text,browser_url,doc_path,display_id,source");
                csvHeaderWritten = true;
            }
        }

        /// <summary>
        /// Closes log files
        /// </summary>
        private void CloseLogFiles()
        {
            ndjsonWriter?.Close();
            csvWriter?.Close();
            ndjsonWriter = null;
            csvWriter = null;
        }

        /// <summary>
        /// Saves a screenshot with UTC timestamp filename
        /// </summary>
        public void SaveScreenshot(Bitmap bitmap, int displayId)
        {
            try
            {
                // Check if we need to create a new folder (date changed)
                string todayFolder = GetTodayFolderName();
                string? currentFolder = currentDateFolder != null ? Path.GetFileName(currentDateFolder) : null;

                if (todayFolder != currentFolder)
                {
                    CloseLogFiles();
                    CreateTodayFolder();
                }

                if (currentDateFolder == null)
                    return;

                // Generate filename: YYYY-MM-DDTHH-MM-SS.mmmZ.png
                DateTime timestamp = DateTime.UtcNow;
                string filename = GetScreenshotFilename(timestamp, displayId);
                string filePath = Path.Combine(currentDateFolder, filename);

                // Save as PNG
                bitmap.Save(filePath, ImageFormat.Png);
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error saving screenshot: {ex.Message}");
            }
        }

        /// <summary>
        /// Generates screenshot filename with ISO-8601 format
        /// </summary>
        private string GetScreenshotFilename(DateTime date, int displayId)
        {
            // Format: YYYY-MM-DDTHH-MM-SS.mmmZ.png
            // ISO8601 gives us: 2025-11-11T14:30:15.123Z
            // We need to replace colons with dashes for filesystem compatibility
            string isoString = date.ToString("yyyy-MM-ddTHH:mm:ss.fffZ", CultureInfo.InvariantCulture);
            string formatted = isoString.Replace(":", "-");

            // Add display ID suffix if not primary display
            string displaySuffix = displayId == 1 ? "" : $"-display{displayId}";

            return $"{formatted}{displaySuffix}.png";
        }

        /// <summary>
        /// Logs a click event to both NDJSON and CSV files
        /// </summary>
        public void LogClick(ClickEvent clickEvent)
        {
            try
            {
                // Check if we need to create a new folder
                string todayFolder = GetTodayFolderName();
                string? currentFolder = currentDateFolder != null ? Path.GetFileName(currentDateFolder) : null;

                if (todayFolder != currentFolder)
                {
                    CloseLogFiles();
                    CreateTodayFolder();
                }

                // Write to NDJSON
                if (ndjsonWriter != null)
                {
                    string ndjson = ToNDJSON(clickEvent);
                    ndjsonWriter.WriteLine(ndjson);
                }

                // Write to CSV
                if (csvWriter != null)
                {
                    string csv = ToCSVRow(clickEvent);
                    csvWriter.WriteLine(csv);
                }
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error logging click: {ex.Message}");
            }
        }

        /// <summary>
        /// Converts click event to NDJSON format
        /// </summary>
        private string ToNDJSON(ClickEvent clickEvent)
        {
            var dict = new Dictionary<string, object>
            {
                { "ts", clickEvent.Timestamp.ToString("yyyy-MM-ddTHH:mm:ss.fffZ", CultureInfo.InvariantCulture) },
                { "x", clickEvent.X },
                { "y", clickEvent.Y },
                { "app", clickEvent.App },
                { "window_title", clickEvent.WindowTitle },
                { "role", clickEvent.Role },
                { "text", clickEvent.Text },
                { "browser_url", clickEvent.BrowserUrl },
                { "doc_path", clickEvent.DocPath },
                { "display_id", clickEvent.DisplayId },
                { "source", clickEvent.Source }
            };

            return JsonConvert.SerializeObject(dict);
        }

        /// <summary>
        /// Converts click event to CSV row
        /// </summary>
        private string ToCSVRow(ClickEvent clickEvent)
        {
            string Escape(string str)
            {
                if (str.Contains(",") || str.Contains("\"") || str.Contains("\n") || str.Contains("\r"))
                {
                    return "\"" + str.Replace("\"", "\"\"") + "\"";
                }
                return str;
            }

            return string.Join(",", new[]
            {
                clickEvent.Timestamp.ToString("yyyy-MM-ddTHH:mm:ss.fffZ", CultureInfo.InvariantCulture),
                clickEvent.X.ToString(),
                clickEvent.Y.ToString(),
                Escape(clickEvent.App),
                Escape(clickEvent.WindowTitle),
                Escape(clickEvent.Role),
                Escape(clickEvent.Text),
                Escape(clickEvent.BrowserUrl),
                Escape(clickEvent.DocPath),
                clickEvent.DisplayId.ToString(),
                clickEvent.Source
            });
        }
    }
}
